### ISSUE TITLE
Reflector getAllAndMerge is not filtering out `undefined` from array

### ISSUE THREAD

**Issue by nllahat** [nestjs/nest#4853]
## Current behavior
When using @Roles on controller level, and leaving some method levels without @Roles the `roles` reflected from the guard contains `undefined` which breaking the `roles` array. 

## Input Code

```ts
@Roles('ADMIN')
@Controller('v1/collections')
export class CollectionsController {
  constructor(private collectionsService: CollectionsService) {}

  @Roles('BASIC')
  @Get(':collectionId')
  async getCollectionById(
    @Param('collectionId') collectionId: string,
  ): Promise<Collection> {
    return this.collectionsService.getCollectionById(collectionId);
  }

  //@Roles('ADMIN') -----> this works fine but 
 //seems redundant if ADMIN role is on controller level 
  @Post()
  @HttpCode(201)
  async createCollection(@Body() data: CreateCollectionDto): Promise<string> {
    return this.collectionsService.createCollection(data);
  }
}

@Injectable()
export class RolesGuard implements CanActivate {
  private logger: Logger = new Logger(RolesGuard.name);
  constructor(private reflector: Reflector) {}

  canActivate(context: ExecutionContext): boolean {
    const roles: string[] = this.reflector.getAllAndMerge<
      string[]
    >('roles', [context.getHandler(), context.getClass()]);

    this.logger.debug(`roles: ${JSON.stringify(roles)}`); // Prints ---> roles: [null,["ADMIN"]]

    if (!roles || !roles.length) {
      return true;
    }

    const request: ExtendedRequest = context.switchToHttp().getRequest();
    const requestor: Requestor = request.requestor;

    if (!requestor || !requestor.roles || !requestor.roles.length) {
      return false;
    }

    const matchingRoles = roles.filter(
      role => requestor.roles.indexOf(role) !== -1,
    );

    const isMatch = !!(matchingRoles && matchingRoles.length);

    return isMatch;
  }
}
```

## Expected behavior
I would expect that `roles` will be: `["ADMIN"]`

## Possible Solution
Maybe `this.reflector.getAllAndMerge` should be:

```ts
    getAllAndMerge(metadataKey, targets) {
        const metadataCollection = this.getAll(metadataKey, targets);
        const filterredUndefined = metadataCollection.filter(element => element !== undefined)
        
        return filterredUndefined.reduce((a, b) => {
            if (Array.isArray(a)) {
                return a.concat(b);
            }
            if (shared_utils_1.isObject(a) && shared_utils_1.isObject(b)) {
                return Object.assign(Object.assign({}, a), b);
            }
            return [a, b];
        });
    }
```

## Environment

<pre><code>
Nest version: 7.0.11
 
For Tooling issues:
- Node version: v10.15.0
- Platform:  Mac
</code></pre>

**Comment by kamilmysliwiec** [comment]
Thanks for reporting. Fixed in 7.1.1

**Comment by thaoula** [comment]
Hi @kamilmysliwiec, 

We recently started getting the following error following an update to the latest version -

TypeError: Reduce of empty array with no initial value
    at Array.reduce (<anonymous>)
    at Reflector.getAllAndMerge (/home/app/node_modules/@nestjs/core/services/reflector.service.js:44:35)

We have a custom guard that is registered as an APP_GUARD. The guard attempts to get the users rights required for a controller or method and checks if the user role has those rights.

We do not decorate every method or or controller so for many requests this will be null.

In version 7.0.2 the following method call did not throw any errors

`let rights = this.reflector.getAllAndMerge<number[]>('rights', [context.getHandler(), context.getClass()]);`

In 7.1.2 I have to wrap the call in a try catch. Do you want me to raise this as a seperate issue?

Regards,
Tarek

**Comment by kamilmysliwiec** [comment]
@thaoula thanks for reporting. Fixed in 7.1.3

**Comment by wayoalamos** [comment]
Hi, im still getting same issue with v - 7.1.5.
